### PR TITLE
Add right click menu functionality to overlays

### PR DESCRIPTION
**Pull request by TheStonedTurtle**
Adds the ability to add right-click menu options to overlays. By default these options requires the `shift` modifier key to be held when right-clicking but this can be changed via the RuneLite config settings.

<details>
<summary>
Code for adding the right click menu options
</summary>

<p>

```java
@Override
public List<MenuOption> buildMenuOptions()
{
	List<MenuOption> options = new ArrayList<>();
	options.add(new MenuOption("Reset", () -> log.info("Reset Option")));
	options.add(new MenuOption("Pause", () -> log.info("Pause Option")));
	options.add(new MenuOption("Something", () -> log.info("Something Option")));
	return options;
}
```

</p>
</details>

### CODE REVIEW

**Comment by TheStonedTurtle** [comment]
I've updated this to use the in-game right click menu and to use the alt modifier instead of shift. I have also moved the resetOverlay functionality to be triggered via the `Reset` option inside the right click menu

This update also requires a Mixin change which exposes `client.openMenu(int x, int y)`
https://gfycat.com/LastingWastefulEarwig

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java [hunk 1]
```diff
@@ -51,4 +56,26 @@ public String getName()
 	{
 		return this.getClass().getSimpleName();
 	}
+
+	boolean hasRightClickMenu()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java]: Remove these 2 methods, the rightClickMenuOptions arraylist is already exposed, so use that? I would maybe wrap it in object and make builder or something, idk

**Comment by Nightfirecat** [comment]
The last state of the overlays being reset in your gif looks like this:

untitled

This should have the two overlays being stacked in the snap corner instead of being on top of each other.

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java [hunk 2]
```diff
@@ -51,4 +56,26 @@ public String getName()
...
+	}
+
+	protected void addRightClickMenuOption(String optionText, Runnable runnable)
+	{
+		rightClickMenuOptions.add(new MenuOption(optionText, runnable));
+	}
+
+	void selectOption(String optionText)
+	{
+		for (MenuOption o : rightClickMenuOptions)
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java]: you can store the list index in Param1

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java [hunk 3]
```diff
@@ -27,21 +27,26 @@
...
 {
+	private String overlayName = null;
 	private Point preferredLocation;
 	private Dimension preferredSize;
 	private OverlayPosition preferredPosition;
 	private Rectangle bounds = new Rectangle();
 	private OverlayPosition position = OverlayPosition.TOP_LEFT;
 	private OverlayPriority priority = OverlayPriority.NONE;
 	private OverlayLayer layer = OverlayLayer.UNDER_WIDGETS;
+	private List<MenuOption> rightClickMenuOptions = new ArrayList<>();
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java]: This can just be `menuOptions`?

**Comment by Nightfirecat** [comment]
I think this basically fixes #1203 as well.

**Comment by TheStonedTurtle** [comment]
It somewhat fixes that issue, it doesn't tell you what plugin it's from just what the overlay filename is. (Unless you override it with `setOverlayName()`)

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 4]
```diff
@@ -494,4 +558,30 @@ private OverlayBounds buildSnapCorners()
...
+		List<MenuOption> options = overlay.getMenuOptions();
+		List<MenuEntry> entries = new ArrayList<>();
+
+		int optionIndex = 0;
+		for (MenuOption option : options)
+		{
+			String name = overlay.getOverlayName() != null ? overlay.getOverlayName() : overlay.getName();
+			final MenuEntry entry = new MenuEntry();
+			entry.setOption(option.getOptionText());
+			entry.setTarget(ColorUtil.colorTag(ColorScheme.BRAND_ORANGE) + name);
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: Is there any reason for using BRAND_ORANGE rather than the default menu target color?

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java [hunk 5]
```diff
@@ -27,21 +27,26 @@
...
 {
+	private String overlayName = null;
 	private Point preferredLocation;
 	private Dimension preferredSize;
 	private OverlayPosition preferredPosition;
 	private Rectangle bounds = new Rectangle();
 	private OverlayPosition position = OverlayPosition.TOP_LEFT;
 	private OverlayPriority priority = OverlayPriority.NONE;
 	private OverlayLayer layer = OverlayLayer.UNDER_WIDGETS;
+	private List<MenuOption> menuOptions = new ArrayList<>();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java]: Why you have both buildMenuOptions and this

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java [hunk 6]
```diff
@@ -27,21 +27,26 @@
...
 import lombok.Getter;
 import lombok.Setter;
 import net.runelite.client.ui.overlay.components.LayoutableRenderableEntity;
+import net.runelite.client.ui.overlay.menu.MenuOption;
 
 @Getter
 @Setter
 public abstract class Overlay implements LayoutableRenderableEntity
 {
+	private String overlayName = null;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java]: You should make this overridable method probably, or use the other method that already exists

Comment (TheStonedTurtle) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java]: This holds the Menu Target name to display, you can `setOverlayName()` inside an overlay to update this value. If this is null it defaults to the existing `getName()`. 

I didn't want to update the original because It is used for config storage.

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java (hunk 4 above)
Comment (TheStonedTurtle) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: Didn't know it existed honestly, i'll change it.

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java (hunk 6 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java]: Storing keys with spaces is fine, so it should be fine and at least there is some default value instead of null.

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java (hunk 5 above)
Comment (TheStonedTurtle) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java]: This stores the menuOptions that were displayed at the time of opening the right-click menu. The MenuEntries store the Overlay Index `Param0` and MenuOption index `Param1` so we can reference the right runnable later on.

Since the results of `buildMenuOptions` could be different than the options displayed in the menu i figured storing it was the best solution.

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java]: Well, that seems wrong. If you have to result to something like this, then I would simply not use the `buildMenuEnries` method at all and simply use just the menuOptions array.

File: runelite-api/src/main/java/net/runelite/api/events/ClientTick.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * An event called once every client tick
+ *
+ * <p>
+ * A client tick is a unit of time used by the RuneScape client.
+ * Each client tick is approximately 20ms or 50 times per second
+ * @see Constants#CLIENT_TICK_LENGTH for more info
+ * <p>
+ */
+@Data
```
Comment (deathbeam) [inline runelite-api/src/main/java/net/runelite/api/events/ClientTick.java]: Remove the annotation? it has no use

File: runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java [hunk 8]
```diff
@@ -261,4 +261,15 @@ default int infoBoxSize()
 	{
 		return 35;
 	}
+
+	@ConfigItem(
+		keyName = "overlayRightClick",
+		name = "Overlay menu hot-key",
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: You should mention shift here (like, Require Shift for overlay menu)

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java [hunk 9]
```diff
@@ -107,6 +110,32 @@ public void onPluginChanged(final PluginChanged event)
 		rebuildOverlayLayers();
 	}
 
+	@Subscribe
+	public void onMenuOptionClicked(MenuOptionClicked event)
+	{
+		if (event.getMenuAction().equals(MenuAction.RUNELITE_OVERLAY))
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java]: Flip this if

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java [hunk 10]
```diff
@@ -107,6 +110,32 @@ public void onPluginChanged(final PluginChanged event)
...
+			int index = event.getActionParam();
+			if (index >= overlays.size())
+			{
+				return;
+			}
+
+			Overlay overlay = overlays.get(index);
+			List<MenuOption> options = overlay.getMenuOptions();
+			int optionIndex = event.getWidgetId();
+			if (optionIndex >= options.size())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java]: You should also check negative value

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java [hunk 11]
```diff
@@ -107,6 +110,32 @@ public void onPluginChanged(final PluginChanged event)
...
 
+	@Subscribe
+	public void onMenuOptionClicked(MenuOptionClicked event)
+	{
+		if (event.getMenuAction().equals(MenuAction.RUNELITE_OVERLAY))
+		{
+			event.consume();
+
+			int index = event.getActionParam();
+			if (index >= overlays.size())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java]: You should also check negative value

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java [hunk 12]
```diff
@@ -107,6 +110,32 @@ public void onPluginChanged(final PluginChanged event)
...
+			Overlay overlay = overlays.get(index);
+			List<MenuOption> options = overlay.getMenuOptions();
+			int optionIndex = event.getWidgetId();
+			if (optionIndex >= options.size())
+			{
+				return;
+			}
+
+			MenuOption option = options.get(optionIndex);
+			option.getRunnable().run();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java]: null-check this

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java [hunk 13]
```diff
@@ -321,4 +350,9 @@ private OverlayPosition loadOverlayPosition(final Overlay overlay)
 		final String locationKey = overlay.getName() + OVERLAY_CONFIG_PREFERRED_POSITION;
 		return configManager.getConfiguration(RUNELITE_CONFIG_GROUP_NAME, locationKey, OverlayPosition.class);
 	}
+
+	int getOverlayIndex(Overlay overlay)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java]: Remove this method, you can already access overlays from outside via getter

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 14]
```diff
@@ -99,6 +111,33 @@ public void onFocusChanged(FocusChanged event)
...
 			inOverlayDraggingMode = false;
+			inRightClickMode = false;
+			menuEntries = null;
+		}
+	}
+
+	@Subscribe
+	protected void onClientTick(ClientTick t)
+	{
+		if (menuEntries != null)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: if (menuEntries == null || client.isMenuOpen())
{
  menuEntries == null;
  return;
}

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 15]
```diff
@@ -99,6 +111,33 @@ public void onFocusChanged(FocusChanged event)
...
+				menuEntries = null;
+				return;
+			}
+
+			client.setMenuEntries(ArrayUtils.addAll(menuEntries, client.getMenuEntries()));
+		}
+	}
+
+	@Subscribe
+	protected void onConfigChanged(ConfigChanged c)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: public

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/menu/MenuOption.java [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.ui.overlay.menu;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
+@AllArgsConstructor
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/menu/MenuOption.java]: Value

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/menu/MenuOption.java [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+public class MenuOption
+{
+	private String optionText;
+	private Runnable runnable;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/menu/MenuOption.java]: callback?

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/menu/MenuOption.java [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+package net.runelite.client.ui.overlay.menu;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
+@AllArgsConstructor
+public class MenuOption
+{
+	private String optionText;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/menu/MenuOption.java]: text

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 19]
```diff
@@ -344,6 +390,13 @@ public MouseEvent mouseReleased(MouseEvent mouseEvent)
 		return mouseEvent;
 	}
 
+	@Override
+	public MouseEvent mouseMoved(MouseEvent mouseEvent)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: Why is this necessary? you can just use client.getMouseCanvasPosition

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java [hunk 20]
```diff
@@ -27,28 +27,33 @@
...
 import lombok.Getter;
 import lombok.Setter;
 import net.runelite.client.ui.overlay.components.LayoutableRenderableEntity;
+import net.runelite.client.ui.overlay.menu.MenuOption;
 
 @Getter
 @Setter
 public abstract class Overlay implements LayoutableRenderableEntity
 {
+	private String name = this.getClass().getSimpleName();
```
[the remark this answers is not in the document]
Comment (TheStonedTurtle) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java]: I thought having this variable and doing `setName()` inside the overlay would be cleaner than having to overriding the existing `getName()` method for the overlay.

Is there a reason why the other way is preferred?

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java (hunk 5 above)
Comment (TheStonedTurtle) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java]: I thought there might be use cases where certain menu options should only be added in specific cases. Just having the array will require manually grabbing and adjusting the array which seems like a lot of duplicate code inside overlays.

I'll make the changes as requested it just seems to be a step backwards IMO.

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/Overlay.java]: Not really, but there is no reason to change it

**Comment by TheStonedTurtle** [comment]
The requested changes have been made.

New way to add Menu options: `getMenuOptions().put("Reset", () -> log.info("Reset Clicked"));`

New way to change overlay name:
```java
@Override
public String getName()
{
	return "Overlay Name";
}
```

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java [hunk 21]
```diff
@@ -113,26 +113,23 @@ public void onPluginChanged(final PluginChanged event)
 	@Subscribe
 	public void onMenuOptionClicked(MenuOptionClicked event)
 	{
-		if (event.getMenuAction().equals(MenuAction.RUNELITE_OVERLAY))
+		if (MenuAction.RUNELITE_OVERLAY.equals(event.getMenuAction()))
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java]: by flipping the if I meant if (!blahblah) return, lol

Comment (TheStonedTurtle) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java]: 🤦‍♂️ Fixed now

File: runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java [hunk 22]
```diff
@@ -261,4 +261,15 @@ default int infoBoxSize()
 	{
 		return 35;
 	}
+
+	@ConfigItem(
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Maybe this should be above infoboxes, under overlay fonts

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java [hunk 23]
```diff
@@ -107,6 +109,30 @@ public void onPluginChanged(final PluginChanged event)
...
+		event.consume();
+
+		int index = event.getActionParam();
+		if (index < 0 || index >= overlays.size())
+		{
+			return;
+		}
+
+		Overlay overlay = overlays.get(index);
+		Runnable r = overlay.getMenuOptions().get(event.getMenuOption());
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java]: as you did not made the menuOptions final, null-check them first here

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 24]
```diff
@@ -227,6 +261,14 @@ public void render(Graphics2D graphics, final OverlayLayer layer)
 					graphics.draw(bounds);
 					graphics.setColor(previous);
 				}
+
+				boolean toggleCheck = inRightClickMode || !runeLiteConfig.overlayRightClick();
+				net.runelite.api.Point mouse = client.getMouseCanvasPosition();
+				mousePosition.setLocation(mouse.getX(), mouse.getY());
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: No need to update mousePosition here, I think?

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 25]
```diff
@@ -356,6 +398,11 @@ public void keyPressed(KeyEvent e)
 		{
 			inOverlayDraggingMode = true;
 		}
+
+		if (runeLiteConfig.overlayRightClick() && e.isShiftDown())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: you should probably just check VK_SHIFT, and make it constant at top of the class

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 26]
```diff
@@ -365,6 +412,12 @@ public void keyReleased(KeyEvent e)
 		{
 			inOverlayDraggingMode = false;
 		}
+
+		if (runeLiteConfig.overlayRightClick() && !e.isShiftDown())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: See my previous comment

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 27]
```diff
@@ -69,6 +77,8 @@
 	private final Point mousePosition = new Point();
 	private Overlay movedOverlay;
 	private boolean inOverlayDraggingMode;
+	private boolean inRightClickMode;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: rename this to e.g inOverlayMenuMode

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 28]
```diff
@@ -494,4 +547,26 @@ private OverlayBounds buildSnapCorners()
 			new Rectangle(rightChatboxPoint, SNAP_CORNER_SIZE),
 			new Rectangle(canvasTopRightPoint, SNAP_CORNER_SIZE));
 	}
+
+	private MenuEntry[] createRightClickMenuEntries(Overlay overlay)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: this can be static right

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 29]
```diff
@@ -494,4 +547,26 @@ private OverlayBounds buildSnapCorners()
...
+	{
+		Set<String> options = overlay.getMenuOptions().keySet();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: final

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 30]
```diff
@@ -494,4 +547,26 @@ private OverlayBounds buildSnapCorners()
...
+		MenuEntry[] entries = new MenuEntry[options.size()];
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: final

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 31]
```diff
@@ -494,4 +547,26 @@ private OverlayBounds buildSnapCorners()
...
+
+		// Add in reverse order so they display correctly in the right-click menu
+		int idx = options.size() - 1;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: final

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 32]
```diff
@@ -494,4 +547,26 @@ private OverlayBounds buildSnapCorners()
...
+		for (String option : options)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: final

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java [hunk 33]
```diff
@@ -107,6 +109,30 @@ public void onPluginChanged(final PluginChanged event)
...
+	public void onMenuOptionClicked(MenuOptionClicked event)
+	{
+		if (!MenuAction.RUNELITE_OVERLAY.equals(event.getMenuAction()))
+		{
+			return;
+		}
+
+		event.consume();
+
+		int index = event.getActionParam();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java]: final

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java [hunk 34]
```diff
@@ -107,6 +109,30 @@ public void onPluginChanged(final PluginChanged event)
...
+		if (index < 0 || index >= overlays.size())
+		{
+			return;
+		}
+
+		Overlay overlay = overlays.get(index);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java]: final

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java (hunk 23 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayManager.java]: final

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 35]
```diff
@@ -227,6 +261,14 @@ public void render(Graphics2D graphics, final OverlayLayer layer)
 					graphics.draw(bounds);
 					graphics.setColor(previous);
 				}
+
+				boolean toggleCheck = inRightClickMode || !runeLiteConfig.overlayRightClick();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: final

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 36]
```diff
@@ -227,6 +261,14 @@ public void render(Graphics2D graphics, final OverlayLayer layer)
...
+				net.runelite.api.Point mouse = client.getMouseCanvasPosition();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: final

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 37]
```diff
@@ -494,4 +547,26 @@ private OverlayBounds buildSnapCorners()
...
+
+		// Add in reverse order so they display correctly in the right-click menu
+		int idx = options.size() - 1;
+		for (String option : options)
+		{
+			final MenuEntry entry = new MenuEntry();
+			entry.setOption(option);
+			entry.setTarget(ColorUtil.colorTag(JagexColors.MENU_TARGET) + overlay.getName());
+			entry.setIdentifier(-1);
+			entry.setParam0(overlayManager.getOverlays().indexOf(overlay));
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: You probably do not need to set this anymore because you are not using it

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java [hunk 38]
```diff
@@ -494,4 +547,26 @@ private OverlayBounds buildSnapCorners()
...
+		Set<String> options = overlay.getMenuOptions().keySet();
+		MenuEntry[] entries = new MenuEntry[options.size()];
+
+		// Add in reverse order so they display correctly in the right-click menu
+		int idx = options.size() - 1;
+		for (String option : options)
+		{
+			final MenuEntry entry = new MenuEntry();
+			entry.setOption(option);
+			entry.setTarget(ColorUtil.colorTag(JagexColors.MENU_TARGET) + overlay.getName());
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: You should use ColorUtil.wrapWithColor so there is terminator

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java (hunk 37 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: you can probably just grab it by name instead

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java (hunk 31 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java]: ah right

**Comment by TheStonedTurtle** [comment]
The requested changes have been implemented.

**Comment by TheStonedTurtle** [comment]
Fixed PR as Client Tick was added in 0d5eac1. Also fixed the commit history.
